lib/lltz_michelson/instruction.ml

```diff
@@ -0,0 +1,499 @@
+  in
+  Config.ok stack
+  @@
+  match n with
+  | 0 -> []
+  | 1 -> [ I.swap ]
+  | n -> [ I.dug_n n ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-DUP, stack is 0-indexed *)
+let dup n stack =
+  (* n is a index into the stack *)
+  let stack =
+    if List.length stack >= n
+    then `Value :: stack
+    else
+      raise_s [%message "Instruction.dup: invalid stack" (stack : SlotStack.t) (n : int)]
+  in
+  Config.ok stack [ I.dup_n (n + 1) ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-SWAP *)
+let swap stack =
+  let stack =
+    match stack with
+    | s0 :: s1 :: slots -> s1 :: s0 :: slots
+    | _ -> raise_s [%message "Instruction.swap: invalid stack" (stack : SlotStack.t)]
+  in
+  Config.ok stack [ I.swap ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-DROP *)
+let drop n stack =
+  (* drop's n elements *)
+  if List.length stack < n
+  then
+    raise_s [%message "Instruction.drop: invalid stack" (stack : SlotStack.t) (n : int)]
+  else (
+    match n with
+    | 0 -> Config.ok stack [] (* noop *)
+    | 1 -> Config.ok (List.tl_exn stack) [ I.drop ] (* drop just one element *)
+    | n -> Config.ok (List.drop stack n) [ I.drop_n n ])
+
+let remove n stack =
+  (* removes element at index n *)
+  let stack =
+    match List.split_n stack n with
+    | left, _nth :: right -> left @ right
+    | _ ->
+      raise_s
+        [%message "Instruction.remove: invalid stack" (stack : SlotStack.t) (n : int)]
+  in
+  Config.ok stack
+  @@
+  match n with
+  | 0 -> I.[ drop ]
+  | 1 -> I.[ swap; drop ]
+  | n -> I.[ dig_n n; drop ]
+
+(* prim m n instr stack: m elements are consumed from the stack, n elements are produced *)
+let prim m n instr stack =
+  let stack =
+    let left, right = List.split_n stack m in
+    (* split stack into left and right at index m *)
+    if not
+         (List.for_all left ~f:(function
+           | `Value -> true
+           | _ -> false))
+    then
+      raise_s
+        [%message
+          "Instruction.prim: invalid stack" (stack : SlotStack.t) (m : int) (n : int)];
+    List.init n ~f:(fun _ -> `Value) @ right
+  in
+  Config.ok stack [ instr ]
+
+let noop stack = Config.ok stack []
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-LOOP *)
+let loop (in_ : t) stack =
+  let stack =
+    match stack with
+    | `Value :: stack -> stack
+    | _ -> raise_s [%message "Instruction.loop" (stack : SlotStack.t)]
+  in
+  let instrs = Config.instructions @@ in_ stack in
+  Config.ok stack [ I.loop instrs ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-LOOP_LEFT *)
+let loop_left (in_ : t) stack =
+  let stack =
+    match stack with
+    | `Value :: _ -> stack
+    | _ -> raise_s [%message "Instruction.loop_left" (stack : SlotStack.t)]
+  in
+  let instrs = Config.instructions @@ in_ stack in
+  Config.ok stack [ I.loop_left instrs ]
+
+(* https://tezos.gitlab.io/michelson-reference/#instr-ITER *)
+let iter (in_ : t) stack =
+  let stack =
```

Comment:
match stack with
`Value :: _ -> let instrs = Config.instructions @@ in_ stack in ...
| _ -> raise_s ...